### DISCUSSION TITLE
How to log loss terms in the DPO calculation?

### DISCUSSION THREAD

**Discussion by mynameismon** [huggingface/trl#3840] [Q&A]
I want to log both the reference chosen rewards and the model chosen rewards, and similarly for the reference rejected rewards and model rejected rewards. Is there any way to do that?

**Comment by ErenAta16** [comment]
DPOTrainer already tracks the rewards you're describing, you don't need to add anything to get them logged.

Under the hood, the "reward" in DPO is defined as beta * (policy_logps - reference_logps), so it's not really separable into a standalone "model reward" and "reference reward" the way you might in PPO. That said, the trainer logs both the combined reward and the raw policy log-probs by default, which is enough to recover everything:

- rewards/chosen and rewards/rejected: beta * (chosen/rejected logratio between policy and reference)
- rewards/margins: rewards/chosen - rewards/rejected
- rewards/accuracies: fraction of the batch where chosen reward beat rejected reward
- logps/chosen and logps/rejected: the policy model's own log-probs, no reference term mixed in

These show up automatically in whatever you passed to report_to (W&B, TensorBoard, etc.) with no extra code.

If you specifically want the reference model's log-probs as their own series (not just baked into the reward), you can get them for free from the two metrics above, since:

ref_chosen_logps = logps/chosen - rewards/chosen / beta

and the same relationship holds for rejected. No trainer subclassing needed, it's just algebra on values that are already being logged.

If you'd rather log the raw reference logps directly instead of deriving them, you'd need to subclass DPOTrainer and override _compute_loss, since that's where ref_chosen_logps/ref_rejected_logps are computed (via compute_ref_log_probs) but not currently pushed into self._metrics.

**Comment by Mahesh-Paul** [comment]
In TRL's `DPOTrainer`, you can log individual loss and margin components by either customizing the evaluation step or overriding `get_batch_loss_metrics`.

### 1. Default Metrics via Callbacks / WandB
`DPOTrainer` computes implicit metrics internally during training and evaluation:
- `rewards/chosen`
- `rewards/rejected`
- `rewards/accuracies`
- `rewards/margins`
- `logps/chosen`
- `logps/rejected`

Make sure you pass your logger in `TrainingArguments`:
```python
training_args = DPOConfig(
    ...,
    logging_steps=10,
    report_to=["wandb"],  # or "tensorboard"
)


2. Custom Logging via Method Override

from trl import DPOTrainer

class CustomDPOTrainer(DPOTrainer):
    def get_batch_loss_metrics(self, model, batch, train_eval="train"):
        loss, metrics = super().get_batch_loss_metrics(model, batch, train_eval=train_eval)
        
        # Log additional customized metrics
        prefix = "eval_" if train_eval == "eval" else ""
        metrics[f"{prefix}custom_loss_term"] = loss.detach().item()
        
        return loss, metrics
